src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BzlmodRepoRuleCreator.java

```diff
@@ -93,4 +99,51 @@ public static Rule createRule(
     }
     return rule;
   }
+
+  public static void validateLabelAttrs(Rule rule, ModuleExtensionId owningExtension)
+      throws EvalException {
+    for (var attribute : rule.getRuleClassObject().getAttributes()) {
+      Object value = rule.getAttr(attribute.getName());
+      Collection<?> toValidate =
+          switch (value) {
+            case List<?> list -> list;
+            case Map<?, ?> map -> map.keySet();
+            case null, default -> singletonList(value);
+          };
+      for (var item : toValidate) {
+        if (!(item instanceof Label label)) {
+          continue;
+        }
+        if (label.getRepository().isVisible()) {
+          continue;
+        }
+        String unprefixedRepoName = label.getName().substring(label.getName().lastIndexOf('~') + 1);
+        RepositoryName owningModuleRepoName = owningExtension.getBzlFileLabel().getRepository();
+        String owningModule;
+        if (owningModuleRepoName.isMain()) {
+          owningModule = "root module";
+        } else {
+          owningModule =
+              String.format(
+                  "module '%s'",
+                  owningModuleRepoName
+                      .getName()
+                      .substring(0, owningModuleRepoName.getName().indexOf('~')));
+        }
+        throw Starlark.errorf(
+            "no repository visible as '@%s', but referenced by label '@%s//%s:%s' in attribute %s"
```

Comment:
I'm tempted to suggest a "deeper" fix here, at the [LabelConverter](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/packages/LabelConverter.java;drc=bc1c1de6b7353b7b31bbd8ce0cde11f7186751d2;l=63) level.

The original reason we didn't outright throw an exception when encountering a label with bad apparent repo names is that it would stop the evaluation of BUILD files. As in, you're trying to build the completely valid `//:foo`, but the unrelated `//:bar` has a dep on a bad label so you can't build your `//:foo` just because it's defined in the same BUILD file.

But as we can see, it's actually desired behavior to fail very eagerly if we're not in a BUILD thread, but a module extension eval thread. The advantage of that is that you actually get the unparsed label text (instead of having to parse it out of the post-conversion label object). The disadvantage is that you don't check things that are actually passed around as Label objects (for example, a tag passes in a bad label, and the extension just forwards it to a repo rule attr). Then again, in that case, the advice produced by this check here would be misleading / outright wrong.

WDYT?

Reply:
> The disadvantage is that you don't check things that are actually passed around as Label objects (for example, a tag passes in a bad label, and the extension just forwards it to a repo rule attr). 

A simpler alternative (less reaching into `Package` and injecting context into a `LabelConverter) that would be useful by itself could be to make `validateLabelAttr` apply to `AttributesValue` instead and then also use it on the attributes created for tags. What do you think about that? It would also work with the `LabelConverter` approach, but may obviate the need to switch to it.

Reply:
Thinking about it a bit more, we indeed don't always have a `LabelConverter` readily available. (I was thinking about the TypeCheckedTag thing where we construct our own `LabelConverter`.) So it would indeed need to reach in to `Package` etc which is very ugly so let's avoid that. I suppose we could revisit this when we get rid of the "each repo is a Rule" thing in Bazel 9.0 or so...